File: src/runtime/tests/capabilities-resolver-test.ts
/**
 * @license
 * Copyright 2020 Google LLC.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * Code distributed by Google as part of this project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */
import {assert} from '../../platform/chai-web.js';
import {Flags} from '../flags.js';
import {VolatileStorageKey} from '../storage/drivers/volatile.js';
import {RamDiskStorageKey} from '../storage/drivers/ramdisk.js';
import {DatabaseStorageKey, MemoryDatabaseStorageKey, PersistentDatabaseStorageKey, MemoryDatabaseStorageKeyFactory, PersistentDatabaseStorageKeyFactory} from '../storage/database-storage-key.js';
import {StorageKey} from '../storage/storage-key.js';
import {ReferenceModeStorageKey} from '../storage/reference-mode-storage-key.js';
import {EntityType, Schema} from '../../types/lib-types.js';
import {CapabilitiesResolver} from '../capabilities-resolver.js';
import {ArcId} from '../id.js';
import {Capabilities, Persistence, Ttl, Shareable, DeletePropagation} from '../capabilities.js';
import {assertThrowsAsync} from '../../testing/test-util.js';
import {Runtime} from '../runtime.js';

describe('Capabilities Resolver', () => {

  let runtime;
  beforeEach(() => {
    runtime = new Runtime();
  });

  type StorageKeyType = typeof VolatileStorageKey|typeof RamDiskStorageKey|typeof DatabaseStorageKey;
  function verifyReferenceModeStorageKey(key: StorageKey, expectedType: StorageKeyType) {
    assert.isTrue(key instanceof ReferenceModeStorageKey);
    const refKey = key as ReferenceModeStorageKey;
    assert.instanceOf(refKey.backingKey, expectedType,
        `Expected ${refKey.backingKey.constructor.name} to be instance of ${expectedType.name}`);
    assert.instanceOf(refKey.storageKey, expectedType,
        `Expected ${refKey.storageKey.constructor.name} to be instance of ${expectedType.name}`);
  }
  const entityType = new EntityType(new Schema(['Thing'], {result: 'Text'}));
  const handleId = 'h0';

  const unspecified = Capabilities.fromAnnotations();
  const inMemory = Capabilities.create([Persistence.inMemory()]);
  const inMemoryWithTtls = Capabilities.create([Persistence.inMemory(), Ttl.days(1)]);
  const inMemoryWithDeleteProp = Capabilities.create([Persistence.inMemory(), new DeletePropagation(true)]);
  const onDisk = Capabilities.create([Persistence.onDisk()]);
  const onDiskWithTtl = Capabilities.create([Persistence.onDisk(), Ttl.minutes(30)]);

  it('fails creating keys with no factories', Flags.withDefaultReferenceMode(async () => {
    const resolver = new CapabilitiesResolver({arcId: ArcId.newForTest('test')});
    // Verify storage keys for none of the capabilities cannot be created.
    await assertThrowsAsync(async () => resolver.createStorageKey(unspecified, entityType, handleId));
    await assertThrowsAsync(async () => resolver.createStorageKey(inMemory, entityType, handleId));
    await assertThrowsAsync(async () => resolver.createStorageKey(inMemoryWithTtls, entityType, handleId));
    await assertThrowsAsync(async () => resolver.createStorageKey(onDisk, entityType, handleId));
    await assertThrowsAsync(async () => resolver.createStorageKey(onDiskWithTtl, entityType, handleId));
  }));

  it('creates volatile keys', Flags.withDefaultReferenceMode(async () => {
    // Register volatile storage key factory.
    // Verify only volatile (in-memory, no ttl) storage key can be created.
    const capabilitiesResolver = runtime.getCapabilitiesResolver(ArcId.newForTest('test'));
    const createKey = capabilitiesResolver.createStorageKey.bind(capabilitiesResolver);
    verifyReferenceModeStorageKey(await createKey(unspecified, entityType, handleId), VolatileStorageKey);
    verifyReferenceModeStorageKey(await createKey(inMemory, entityType, handleId), VolatileStorageKey);
    await assertThrowsAsync(async () => createKey(inMemoryWithTtls, entityType, handleId));
    await assertThrowsAsync(async () => createKey(onDisk, entityType, handleId));
    await assertThrowsAsync(async () => createKey(onDiskWithTtl, entityType, handleId));
    await assertThrowsAsync(async () => createKey(inMemoryWithDeleteProp, entityType, handleId));
  }));

  it('creates keys with db only factories', Flags.withDefaultReferenceMode(async () => {
    const resolver = new CapabilitiesResolver({
      arcId: ArcId.newForTest('test'),
      factories: [new PersistentDatabaseStorageKeyFactory(), new MemoryDatabaseStorageKeyFactory()]
    });
    const createKey = resolver.createStorageKey.bind(resolver);
    verifyReferenceModeStorageKey(await createKey(unspecified, entityType, handleId), MemoryDatabaseStorageKey);
    verifyReferenceModeStorageKey(await createKey(inMemory, entityType, handleId), MemoryDatabaseStorageKey);
    verifyReferenceModeStorageKey(await createKey(inMemoryWithTtls, entityType, handleId), MemoryDatabaseStorageKey);
    verifyReferenceModeStorageKey(await createKey(inMemoryWithDeleteProp, entityType, handleId), MemoryDatabaseStorageKey);
    verifyReferenceModeStorageKey(await createKey(onDisk, entityType, handleId), PersistentDatabaseStorageKey);
    verifyReferenceModeStorageKey(await createKey(onDiskWithTtl, entityType, handleId), PersistentDatabaseStorageKey);
  }));

  it('creates keys with volatile and db factories', Flags.withDefaultReferenceMode(async () => {
    // Register database storage key factories. Verify all storage keys created as expected.
    DatabaseStorageKey.register(runtime);
    const resolver = runtime.getCapabilitiesResolver(ArcId.newForTest('test'));
    const verify = async (capabilities, type, handleId, keyClass) => verifyReferenceModeStorageKey(
      await resolver.createStorageKey(capabilities, type, handleId), keyClass);
    await verify(unspecified, entityType, handleId, VolatileStorageKey);
    await verify(Capabilities.create([new Shareable(false)]), entityType, handleId, VolatileStorageKey);
    await verify(Capabilities.create([new Shareable(true)]), entityType, handleId, RamDiskStorageKey);
    await verify(inMemory, entityType, handleId, VolatileStorageKey);
    await verify(inMemoryWithTtls, entityType, handleId, MemoryDatabaseStorageKey);
    await verify(onDisk, entityType, handleId, PersistentDatabaseStorageKey);
    await verify(onDiskWithTtl, entityType, handleId, PersistentDatabaseStorageKey);
  }));

  it('creates keys with custom factory', Flags.withDefaultReferenceMode(async () => {
    const resolver = runtime.getCapabilitiesResolver(
        ArcId.newForTest('test'), [new MemoryDatabaseStorageKeyFactory()]);
    verifyReferenceModeStorageKey(await resolver.createStorageKey(unspecified, entityType, handleId), VolatileStorageKey);
    verifyReferenceModeStorageKey(await resolver.createStorageKey(inMemory, entityType, handleId), VolatileStorageKey);
    verifyReferenceModeStorageKey(await resolver.createStorageKey(inMemoryWithTtls, entityType, handleId), MemoryDatabaseStorageKey);
    await assertThrowsAsync(async () => resolver.createStorageKey(onDisk, entityType, handleId));
    await assertThrowsAsync(async () => resolver.createStorageKey(onDiskWithTtl, entityType, handleId));
  }));

  it('creates keys for recipe with volatile and db factories', Flags.withDefaultReferenceMode(async () => {
    DatabaseStorageKey.register(runtime);
    const manifestStr = `
        recipe
          h0: create
          h1: create @ttl('2h')
          h2: create @persistent
          h3: create @persistent @ttl('30m')
          h4: create @queryable @ttl('30m')
    `;
    const recipe = (await runtime.parse(manifestStr)).recipes[0];
    const resolver = runtime.getCapabilitiesResolver(ArcId.newForTest('test'));

    verifyReferenceModeStorageKey(await resolver.createStorageKey(
        recipe.handles[0].capabilities, entityType, handleId), VolatileStorageKey);
    verifyReferenceModeStorageKey(await resolver.createStorageKey(
        recipe.handles[1].capabilities, entityType, handleId), MemoryDatabaseStorageKey);
    verifyReferenceModeStorageKey(await resolver.createStorageKey(
        recipe.handles[2].capabilities, entityType, handleId), PersistentDatabaseStorageKey);
    verifyReferenceModeStorageKey(await resolver.createStorageKey(
        recipe.handles[3].capabilities, entityType, handleId), PersistentDatabaseStorageKey);
    verifyReferenceModeStorageKey(await resolver.createStorageKey(
        recipe.handles[4].capabilities, entityType, handleId), MemoryDatabaseStorageKey);
  }));
});

File: javatests/arcs/core/storage/CapabilitiesResolverTest.kt
/*
 * Copyright 2020 Google LLC.
 *
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 *
 * Code distributed by Google as part of this project is also subject to an additional IP rights
 * grant found at
 * http://polymer.github.io/PATENTS.txt
 */

package arcs.core.storage

import arcs.core.common.ArcId
import arcs.core.data.Capabilities
import arcs.core.data.Capability.Persistence
import arcs.core.data.Capability.Shareable
import arcs.core.data.Capability.Ttl
import arcs.core.data.EntityType
import arcs.core.data.FieldType
import arcs.core.data.ReferenceType
import arcs.core.data.Schema
import arcs.core.data.SchemaFields
import arcs.core.data.SchemaName
import arcs.core.storage.CapabilitiesResolver.Options
import arcs.core.storage.api.DriverAndKeyConfigurator
import arcs.core.storage.keys.DatabaseStorageKey
import arcs.core.storage.keys.RamDiskStorageKey
import arcs.core.storage.keys.VolatileStorageKey
import arcs.core.storage.referencemode.ReferenceModeStorageKey
import com.google.common.truth.Truth.assertThat
import kotlin.test.assertFailsWith
import org.junit.After
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

/** Tests for [CapabilitiesResolver]. */
@RunWith(JUnit4::class)
class CapabilitiesResolverTest {
  @After
  fun tearDown() {
    CapabilitiesResolver.reset()
  }

  private inline fun <reified T : StorageKey> verifyStorageKey(key: StorageKey) {
    assertThat(key).isInstanceOf(ReferenceModeStorageKey::class.java)
    assertThat((key as ReferenceModeStorageKey).backingKey is T).isTrue()
    assertThat(key.storageKey is T).isTrue()
  }

  @Test
  fun registerKeyCreator_nonUnique_fails() {
    CapabilitiesResolver.registerStorageKeyFactory(VolatileStorageKey.VolatileStorageKeyFactory())
    assertFailsWith<IllegalArgumentException> {
      CapabilitiesResolver.registerStorageKeyFactory(VolatileStorageKey.VolatileStorageKeyFactory())
    }
  }

  @Test
  fun createStorageKey_unsupported_fails() {
    val resolver = CapabilitiesResolver(Options(ArcId.newForTest("test")))
    // Verify storage keys for none of the capabilities cannot be created.
    assertFailsWith<IllegalArgumentException> {
      resolver.createStorageKey(UNSPECIFIED, ENTITY_TYPE, HANDLE_ID)
    }
    assertFailsWith<IllegalArgumentException> {
      resolver.createStorageKey(IN_MEMORY, ENTITY_TYPE, HANDLE_ID)
    }
    assertFailsWith<IllegalArgumentException> {
      resolver.createStorageKey(IN_MEMORY_WITH_TTLS, ENTITY_TYPE, HANDLE_ID)
    }
    assertFailsWith<IllegalArgumentException> {
      resolver.createStorageKey(ON_DISK, ENTITY_TYPE, HANDLE_ID)
    }
    assertFailsWith<IllegalArgumentException> {
      resolver.createStorageKey(ON_DISK_WITH_TTL, ENTITY_TYPE, HANDLE_ID)
    }
  }

  @Test
  fun createStorageKey_volatileRegistered_onlyVolatileCreated() {
    // Register volatile storage key factory.
    CapabilitiesResolver.registerStorageKeyFactory(VolatileStorageKey.VolatileStorageKeyFactory())
    val resolver = CapabilitiesResolver(Options(ArcId.newForTest("test")))
    // Verify only volatile (in-memory, no ttl) storage key can be created.
    verifyStorageKey<VolatileStorageKey>(
      resolver.createStorageKey(UNSPECIFIED, ENTITY_TYPE, HANDLE_ID)
    )
    verifyStorageKey<VolatileStorageKey>(
      resolver.createStorageKey(IN_MEMORY, ENTITY_TYPE, HANDLE_ID)
    )
    assertFailsWith<IllegalArgumentException> {
      resolver.createStorageKey(IN_MEMORY_WITH_TTLS, ENTITY_TYPE, HANDLE_ID)
    }
    assertFailsWith<IllegalArgumentException> {
      resolver.createStorageKey(ON_DISK, ENTITY_TYPE, HANDLE_ID)
    }
    assertFailsWith<IllegalArgumentException> {
      resolver.createStorageKey(ON_DISK_WITH_TTL, ENTITY_TYPE, HANDLE_ID)
    }
  }

  @Test
  fun createStorageKey_database_success() {
    CapabilitiesResolver.registerStorageKeyFactory(DatabaseStorageKey.Memory.Factory())
    CapabilitiesResolver.registerStorageKeyFactory(DatabaseStorageKey.Persistent.Factory())
    val resolver = CapabilitiesResolver(Options(ArcId.newForTest("test")))
    verifyStorageKey<DatabaseStorageKey.Memory>(
      resolver.createStorageKey(UNSPECIFIED, ENTITY_TYPE, HANDLE_ID)
    )
    verifyStorageKey<DatabaseStorageKey.Memory>(
      resolver.createStorageKey(IN_MEMORY, ENTITY_TYPE, HANDLE_ID)
    )
    verifyStorageKey<DatabaseStorageKey.Memory>(
      resolver.createStorageKey(IN_MEMORY_WITH_TTLS, ENTITY_TYPE, HANDLE_ID)
    )
    verifyStorageKey<DatabaseStorageKey.Persistent>(
      resolver.createStorageKey(ON_DISK, ENTITY_TYPE, HANDLE_ID)
    )
    verifyStorageKey<DatabaseStorageKey.Persistent>(
      resolver.createStorageKey(ON_DISK_WITH_TTL, ENTITY_TYPE, HANDLE_ID)
    )
  }

  @Test
  fun createStorageKey_registerAllKeys_success() {
    DriverAndKeyConfigurator.configureKeyParsersAndFactories()
    val resolver = CapabilitiesResolver(Options(ArcId.newForTest("test")))
    verifyStorageKey<VolatileStorageKey>(
      resolver.createStorageKey(UNSPECIFIED, ENTITY_TYPE, HANDLE_ID)
    )

    verifyStorageKey<VolatileStorageKey>(
      resolver.createStorageKey(Capabilities(Shareable(false)), ENTITY_TYPE, HANDLE_ID)
    )
    verifyStorageKey<RamDiskStorageKey>(
      resolver.createStorageKey(Capabilities(Shareable(true)), ENTITY_TYPE, HANDLE_ID)
    )
    verifyStorageKey<VolatileStorageKey>(
      resolver.createStorageKey(IN_MEMORY, ENTITY_TYPE, HANDLE_ID)
    )
    verifyStorageKey<DatabaseStorageKey.Memory>(
      resolver.createStorageKey(IN_MEMORY_WITH_TTLS, ENTITY_TYPE, HANDLE_ID)
    )
    verifyStorageKey<DatabaseStorageKey.Persistent>(
      resolver.createStorageKey(ON_DISK, ENTITY_TYPE, HANDLE_ID)
    )
    verifyStorageKey<DatabaseStorageKey.Persistent>(
      resolver.createStorageKey(ON_DISK_WITH_TTL, ENTITY_TYPE, HANDLE_ID)
    )
  }

  @Test
  fun createStorageKey_inMemoryCapabilitiesInMemoryFactories_success() {
    CapabilitiesResolver.registerStorageKeyFactory(VolatileStorageKey.VolatileStorageKeyFactory())
    val resolver = CapabilitiesResolver(
      Options(ArcId.newForTest("test")),
      listOf(DatabaseStorageKey.Memory.Factory())
    )

    verifyStorageKey<VolatileStorageKey>(
      resolver.createStorageKey(UNSPECIFIED, ENTITY_TYPE, HANDLE_ID)
    )
    verifyStorageKey<VolatileStorageKey>(
      resolver.createStorageKey(IN_MEMORY, ENTITY_TYPE, HANDLE_ID)
    )
    verifyStorageKey<DatabaseStorageKey.Memory>(
      resolver.createStorageKey(IN_MEMORY_WITH_TTLS, ENTITY_TYPE, HANDLE_ID)
    )
  }

  @Test
  fun createStorageKey_dbCapabilitiesInMemoryFactories_fails() {
    CapabilitiesResolver.registerStorageKeyFactory(VolatileStorageKey.VolatileStorageKeyFactory())
    val resolver = CapabilitiesResolver(
      Options(ArcId.newForTest("test")),
      listOf(DatabaseStorageKey.Memory.Factory())
    )

    assertFailsWith<IllegalArgumentException> {
      resolver.createStorageKey(ON_DISK, ENTITY_TYPE, HANDLE_ID)
    }
    assertFailsWith<IllegalArgumentException> {
      resolver.createStorageKey(ON_DISK_WITH_TTL, ENTITY_TYPE, HANDLE_ID)
    }
  }

  @Test
  fun createStorageKey_inMemoryCapabilitiesInMemoryCustomFactory_success() {
    val resolver = CapabilitiesResolver(
      Options(ArcId.newForTest("test")),
      listOf(DatabaseStorageKey.Memory.Factory())
    )

    verifyStorageKey<DatabaseStorageKey.Memory>(
      resolver.createStorageKey(UNSPECIFIED, ENTITY_TYPE, HANDLE_ID)
    )
    verifyStorageKey<DatabaseStorageKey.Memory>(
      resolver.createStorageKey(IN_MEMORY, ENTITY_TYPE, HANDLE_ID)
    )
    verifyStorageKey<DatabaseStorageKey.Memory>(
      resolver.createStorageKey(IN_MEMORY_WITH_TTLS, ENTITY_TYPE, HANDLE_ID)
    )
    assertFailsWith<IllegalArgumentException> {
      resolver.createStorageKey(ON_DISK, ENTITY_TYPE, HANDLE_ID)
    }
    assertFailsWith<IllegalArgumentException> {
      resolver.createStorageKey(ON_DISK_WITH_TTL, ENTITY_TYPE, HANDLE_ID)
    }
  }

  @Test
  fun createStorageKey_dbCapabilitiesInCustomFactory_fails() {
    val resolver = CapabilitiesResolver(
      Options(ArcId.newForTest("test")),
      listOf(DatabaseStorageKey.Memory.Factory())
    )
    assertFailsWith<IllegalArgumentException> {
      resolver.createStorageKey(ON_DISK, ENTITY_TYPE, HANDLE_ID)
    }
    assertFailsWith<IllegalArgumentException> {
      resolver.createStorageKey(ON_DISK_WITH_TTL, ENTITY_TYPE, HANDLE_ID)
    }
  }

  @Test
  fun createStorageKey_referenceType_success() {
    DriverAndKeyConfigurator.configureKeyParsersAndFactories()
    val resolver = CapabilitiesResolver(Options(ArcId.newForTest("test")))
    assertThat(resolver.createStorageKey(UNSPECIFIED, THING_REFERENCE_TYPE, HANDLE_ID))
      .isInstanceOf(VolatileStorageKey::class.java)
    assertThat(
      resolver.createStorageKey(Capabilities(Shareable(false)), THING_REFERENCE_TYPE, HANDLE_ID)
    ).isInstanceOf(VolatileStorageKey::class.java)
    assertThat(
      resolver.createStorageKey(Capabilities(Shareable(true)), THING_REFERENCE_TYPE, HANDLE_ID)
    ).isInstanceOf(RamDiskStorageKey::class.java)
    assertThat(resolver.createStorageKey(IN_MEMORY, THING_REFERENCE_TYPE, HANDLE_ID))
      .isInstanceOf(VolatileStorageKey::class.java)
    assertThat(resolver.createStorageKey(IN_MEMORY_WITH_TTLS, THING_REFERENCE_TYPE, HANDLE_ID))
      .isInstanceOf(DatabaseStorageKey.Memory::class.java)
    assertThat(resolver.createStorageKey(ON_DISK, THING_REFERENCE_TYPE, HANDLE_ID))
      .isInstanceOf(DatabaseStorageKey.Persistent::class.java)
    assertThat(resolver.createStorageKey(ON_DISK_WITH_TTL, THING_REFERENCE_TYPE, HANDLE_ID))
      .isInstanceOf(DatabaseStorageKey.Persistent::class.java)
  }

  @Test
  fun createStorageKey_referenceTypeIncompatibleKeys_fail() {
    val resolver = CapabilitiesResolver(Options(ArcId.newForTest("test")), listOf(FAKE_FACTORY))
    val e = assertFailsWith<IllegalStateException> {
      resolver.createStorageKey(UNSPECIFIED, ENTITY_TYPE, HANDLE_ID)
    }
    assertThat(e).hasMessageThat().isEqualTo("Backing and containers keys must use same protocol")
  }

  @Test
  fun ctor_nonUniqueFactoryProtocols_fail() {
    val e = assertFailsWith<IllegalArgumentException> {
      CapabilitiesResolver(Options(ArcId.newForTest("test")), listOf(FAKE_FACTORY, FAKE_FACTORY))
    }
    assertThat(e).hasMessageThat().isEqualTo(
      "Storage keys protocol must be unique, but was: [test, test]."
    )
  }

  companion object {
    private val ENTITY_TYPE = EntityType(
      Schema(
        setOf(SchemaName("Thing")),
        SchemaFields(mapOf("name" to FieldType.Text), emptyMap()),
        "42"
      )
    )
    private val THING_REFERENCE_TYPE = ReferenceType(ENTITY_TYPE)
    private val HANDLE_ID = "h0"

    private val UNSPECIFIED = Capabilities()
    private val IN_MEMORY = Capabilities(Persistence.IN_MEMORY)
    private val IN_MEMORY_WITH_TTLS = Capabilities(listOf(Persistence.IN_MEMORY, Ttl.Days(1)))
    private val ON_DISK = Capabilities(Persistence.ON_DISK)
    private val ON_DISK_WITH_TTL = Capabilities(listOf(Persistence.ON_DISK, Ttl.Days(1)))

    val FAKE_FACTORY = object : StorageKeyFactory(protocol = "test", capabilities = IN_MEMORY) {
      override fun create(options: StorageKeyOptions): StorageKey {
        return if (options is StorageKeyFactory.ContainerStorageKeyOptions) {
          DatabaseStorageKey.Persistent(options.location, options.entitySchema.hash)
        } else DatabaseStorageKey.Memory(options.location, options.entitySchema.hash)
      }
    }
  }
}
